src/Bicep.Core/TypeSystem/DiscriminatedObjectType.cs

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Bicep.Core.TypeSystem
+{
+    public class DiscriminatedObjectType : TypeSymbol
```

Comment:
>TypeSymbol [](start = 43, length = 10)

Should this be derived from object type?

Reply:
Conceptually in terms of the type system that is exposed to users, yes. However it's a little tricky to try and match that hierarchy in the C# code, and I'm not convinced there's a good reason to do it.

Reply:
Yeah I think you're right.